core/client.go

```diff
@@ -0,0 +1,482 @@
+package core
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+	"unicode/utf8"
+)
+
+var (
+	JsonCheck       = regexp.MustCompile(`(?i:(?:application|problem)/(?:hal\+)?json)`)
+	queryParamSplit = regexp.MustCompile(`(^|&)([^&]+)`)
+	queryDescape    = strings.NewReplacer("%5B", "[", "%5D", "]")
+)
+
+// APIClient In most cases there should be only one, shared, APIClient.
+type APIClient struct {
+	cfg    *Configuration
+	Common Service // Reuse a single struct instead of allocating one for each service on the heap.
+}
+
+type Service struct {
+	Client *APIClient
+}
+
+// NewAPIClient creates a new API client. Requires a userAgent string describing your application.
+// optionally a custom http.Client to allow for advanced features such as caching.
+func NewAPIClient(cfg *Configuration) *APIClient {
+	if cfg.HTTPClient == nil {
+		cfg.HTTPClient = http.DefaultClient
+	}
+
+	c := &APIClient{}
```

Comment:
Nitpick; just like this more.

```suggestion
	c := new(APIClient)
```